### PR TITLE
Release 2.1.0-rc.8

### PR DESCRIPTION
**Pull request by jonathan-casarrubias**
#### What type of pull request are you creating?
- [X] Bug Fix
- [X] Enhancement

#### How many unit test did you write for this pull request?

Write a reason if none (e.g just fixed a typo):

#### Please add a description for your pull request:

- Milestone Details: https://github.com/mean-expert-official/loopback-sdk-builder/milestone/30?closed=1

- Fix: https://github.com/mean-expert-official/loopback-sdk-builder/issues/297
- Fix: https://github.com/mean-expert-official/loopback-sdk-builder/issues/296
- Fix: https://github.com/mean-expert-official/loopback-sdk-builder/issues/295
- Fix: https://github.com/mean-expert-official/loopback-sdk-builder/issues/292
- Fix: https://github.com/mean-expert-official/loopback-sdk-builder/issues/290
- Fix: https://github.com/mean-expert-official/loopback-sdk-builder/issues/288
- Fix: https://github.com/mean-expert-official/loopback-sdk-builder/issues/285
- Fix: https://github.com/mean-expert-official/loopback-sdk-builder/issues/284

### CODE REVIEW

File: lib/angular2/shared/models/flref.ejs [hunk 1]
```diff
@@ -203,10 +257,16 @@ export class FireLoopRef<T> {
...
+    this.socket.on(`${ this.model.getModelName() }.value.result.${ this.id }`, (res: any) => {
+      if (res.error) {
+        subject.error(res);
+      } else {
+        subject.next(res);
+      }
+    });
+    // This event listener will be wiped within socket.connections
+    this.socket.sharedObservables.sharedOnDisconnect.subscribe(() => subject.complete());
+    return subject.asObservable().catch((error: any) => Observable.throw(error));
```
Comment (mike-packetwerk) [inline lib/angular2/shared/models/flref.ejs]: Doing a catch and immediately a throw seems to me to be a noop. Does this actually make a difference?

Comment (jonathan-casarrubias) [inline lib/angular2/shared/models/flref.ejs]: yes a lot. Subjects are being destroyed on errors, catching the error avoids it

File: lib/angular2/shared/sockets/connections.ts [hunk 2]
```diff
@@ -2,85 +2,156 @@
...
+   * @description
+   * This method will wrap the original "on" method and run it within the Angular Zone
+   **/
+  public on(event: string, handler: Function): void {
+    this.socket.on(event, (data: any) => this.zone.run(() => handler(data)));
+  }
+  /**
+   * @method emit
+   * @description
+   * This method will wrap the original "on" method and run it within the Angular Zone
```
Comment (mike-packetwerk) [inline lib/angular2/shared/sockets/connections.ts]: this comment is a duplicate of the on method

File: lib/angular2/shared/sockets/connections.ts [hunk 3]
```diff
@@ -2,85 +2,156 @@
...
+   **/
+  public emit(event: string, data: any): void {
+    this.socket.emit(event, data);
+  }
+  /**
+   * @method removeListener
+   * @description
+   * This method will wrap the original "on" method and run it within the Angular Zone
```
Comment (mike-packetwerk) [inline lib/angular2/shared/sockets/connections.ts]: this comment is a duplicate of the on method

File: lib/angular2/shared/sockets/connections.ts [hunk 4]
```diff
@@ -2,85 +2,156 @@
...
+   * @method heartbeater
+   * @description
+   * This will keep the connection as active, even when users are not sending
+   * data, this avoids disconnection because of a connection not being used.
+   **/
+  private heartbeater(): void {
     let heartbeater: any = setInterval(() => {
-      if (this.connections[url]) {
-        this.connections[url].emit('lb-ping');
+      if (this.socket && this.socket.connected) {
```
Comment (mike-packetwerk) [inline lib/angular2/shared/sockets/connections.ts]: could be replaced with `this.isConnected()`

File: lib/angular2/shared/sockets/connections.ts (hunk 2 above)
Comment (jonathan-casarrubias) [inline lib/angular2/shared/sockets/connections.ts]: Cool thanks Ill fix it once I get a charger for my macbook, mine died yesterday :(

File: lib/angular2/shared/sockets/connections.ts (hunk 4 above)
Comment (jonathan-casarrubias) [inline lib/angular2/shared/sockets/connections.ts]: Could be, but would not that add an aditional fn call? making the stack of calls larger. Not big deal just wanted to avoid that

Thoughts?

Comment (mike-packetwerk) [inline lib/angular2/shared/sockets/connections.ts]: I'd avoid duplicate code, especial when "if" logic is concerned. The concept of connected should be defined once, and referenced by everyone that wants to know. Always calling `isConnected()` to know that state means when/if you change the definition of "is connected" you change it in exactly one place.

File: lib/angular2/shared/models/flref.ejs (hunk 1 above)
Comment (mike-packetwerk) [inline lib/angular2/shared/models/flref.ejs]: Hmm, I don't understand your answer, but ok. I'll try the code with and without the catch-throw and see what happens.

File: lib/angular2/shared/sockets/connections.ts (hunk 4 above)
Comment (jonathan-casarrubias) [inline lib/angular2/shared/sockets/connections.ts]: Fair enough, I will update that for next version.

File: lib/angular2/shared/models/flref.ejs (hunk 1 above)
Comment (jonathan-casarrubias) [inline lib/angular2/shared/models/flref.ejs]: This is  a rxjs thing. Subejcts are terminated on errors and on completed. 

This is a realtime platform and we dont want to dtop a channel.of communication just because you got an error.

If you simply use the subject.error method, your subscribers wont receive any further message from the server using this reference.

I hope this clarifies your questions
